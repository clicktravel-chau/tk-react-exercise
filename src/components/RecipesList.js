import React, { useState, useEffect } from 'react';
import RecipeService from '../services/RecipeService';
import { Link } from 'react-router-dom';

const RecipesList = () => {
  const [recipes, setRecipes] = useState([]);
  const [currentRecipe, setCurrentRecipe] = useState(null);
  const [currentIndex, setCurrentIndex] = useState(-1);
  const [searchName, setSearchName] = useState("");

  useEffect(() => {
    retrieveRecipes();
  }, []);

  const onChangeSearchName = event => {
    const searchName = event.target.value;
    setSearchName(searchName);
  };

  const retrieveRecipes = () => {
    RecipeService.getAll()
      .then(response => {
        setRecipes(response.data);
        console.log(response.data);
      })
      .catch(e => {
        console.log(e);
      });
  };

  const setActiveRecipe = (recipe, index) => {
    setCurrentRecipe(recipe);
    setCurrentIndex(index);
  };

  const findByName = () => {
    RecipeService.findByName(searchName)
      .then(response => {
        setRecipes(response.data);
        console.log(response.data);
      })
      .catch(e => {
        console.log(e);
      });
  };

  return (
    <div className="list row">
      <div className="col-md-8">
        <div className="input-group mb-3">
          <input
            type="text"
            className="form-control"
            placeholder="Search by name"
            value={searchName}
            onChange={onChangeSearchName}
            data-testid="recipes-search-bar"
          />
          <div className="input-group-append">
            <button
              className="btn btn-outline-secondary"
              type="button"
              onClick={findByName}
              data-testid="recipes-search-btn"
            >Search
            </button>
          </div>
        </div>
      </div>
      <div className="col-md-6">
        <h4 data-testid="recipes-header">Recipes List</h4>

        <ul data-testid="recipes-listgroup" className="list-group">
          {
            recipes && recipes.map((recipe, index) => (
              <li
                data-testid={"recipe-" + recipe.id}
                className={
                  "list-group-item " + (index === currentIndex ? "active" : "")
                }
                onClick={() => setActiveRecipe(recipe, index)}
                key={index}
              >
                {recipe.name}
              </li>
            ))
          }
        </ul>
      </div>
      <div className="col-md-6">
        {currentRecipe ? (
          <div data-testid="recipe-details">
            <h4>Recipe</h4>
            <div data-testid="recipe-name">
              <label>
                <strong>Name:</strong>
              </label>{" "}
              {currentRecipe.name}
            </div>
            <div data-testid="recipe-desc">
              <label>
                <strong>Description:</strong>
              </label>{" "}
              {currentRecipe.description}
            </div>
            <div data-testid="recipe-ing">
              <label>
                <strong>Ingredients:</strong>
              </label>{" "}
              <ul>
                {currentRecipe.ingredients.map(ingredient =>
                  <li key={ingredient.id}>{ingredient.name}</li>
                )}
              </ul>
            </div>

            <Link
              to={"/recipe/recipes/" + currentRecipe.id}
              className="btn btn-secondary"
              data-testid="recipe-edit"
            >
              Edit
            </Link>
          </div>
        ) : (
          <div>
            <br />
            <p data-testid="empty-recipe-details">Please click on a Recipe</p>
          </div>
        )}
      </div>
    </div>
  );
};

export default RecipesList;
