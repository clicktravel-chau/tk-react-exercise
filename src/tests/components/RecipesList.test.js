import React from 'react';
import { render, cleanup, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import axiosMock from 'axios';
import RecipesList from '../../components/RecipesList';
import {BrowserRouter} from 'react-router-dom';

describe('RecipesList', () => {
  afterEach(cleanup);

  const response = {
    data: [
      {
        id: 1,
        name: 'Pizza',
        description: 'Bake in the oven',
        ingredients: [{ id: 3, name: 'cheese' }],
      },
      {
        id: 2,
        name: 'Spaghetti pizza',
        description: 'Put it in the oven',
        ingredients: [{ id: 4, name: 'dough' }]
      },
    ],
  }

  it('renders', async () => {
    axiosMock.get.mockResolvedValue(response);
    const { asFragment, getByTestId } = render(<RecipesList />);
    expect(asFragment()).toMatchSnapshot();
    await waitFor(() => expect(getByTestId('recipes-listgroup')).toBeVisible());
  });

  describe('when a list of recipes are being retrieved', () => {
    it('should render a list of recipes after retrieving them and render the search bar',
      async () => {
      // Given
      axiosMock.get.mockResolvedValue(response);

      // When
      const { getByTestId } = render(<RecipesList />);

      // Then
      expect(getByTestId('recipes-search-bar')).toBeVisible();
      expect(getByTestId('recipes-search-btn')).toBeVisible();
      expect(getByTestId('recipes-header')).toBeVisible();
      await waitFor(() => expect(getByTestId('recipe-1')).toHaveTextContent('Pizza'));
      await waitFor(() => expect(getByTestId('recipe-2')).toHaveTextContent('Spaghetti pizza'));
      expect(getByTestId('empty-recipe-details')).toBeVisible();
      expect(axiosMock.get).toHaveBeenCalledTimes(1);
    });

    it('should show an empty list', async () => {
      // Given
      axiosMock.get.mockResolvedValue({ data: [] });

      // When
      const { getByTestId } = render(<RecipesList />);

      // Then
      await waitFor(() => expect(getByTestId('recipes-listgroup')).toBeEmptyDOMElement());
    });

    it('should render the selected recipe from the recipe list', async () => {
      // Given
      axiosMock.get.mockResolvedValue(response);

      // When
      // Need to wrap in BrowserRouter otherwise test is hit by an error,
      // You should not use <Link> outside a <Router>
      const { getByTestId } = render(<BrowserRouter><RecipesList /></BrowserRouter>);

      // Then
      await waitFor(() => expect(getByTestId('recipes-listgroup')).toBeVisible());
      fireEvent.click(getByTestId('recipe-1'));
      await waitFor(() => expect(getByTestId('recipe-details')).toBeVisible());
      expect(getByTestId('recipe-name')).toHaveTextContent('Pizza');
      expect(getByTestId('recipe-desc')).toHaveTextContent('Bake in the oven');
      expect(getByTestId('recipe-ing')).toHaveTextContent('cheese');
      expect(getByTestId('recipe-edit')).toBeVisible();
    });
  });
});
