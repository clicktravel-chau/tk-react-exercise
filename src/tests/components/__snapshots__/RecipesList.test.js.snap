// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipesList renders 1`] = `
<DocumentFragment>
  <div
    class="list row"
  >
    <div
      class="col-md-8"
    >
      <div
        class="input-group mb-3"
      >
        <input
          class="form-control"
          data-testid="recipes-search-bar"
          placeholder="Search by name"
          type="text"
          value=""
        />
        <div
          class="input-group-append"
        >
          <button
            class="btn btn-outline-secondary"
            data-testid="recipes-search-btn"
            type="button"
          >
            Search
          </button>
        </div>
      </div>
    </div>
    <div
      class="col-md-6"
    >
      <h4
        data-testid="recipes-header"
      >
        Recipes List
      </h4>
      <ul
        class="list-group"
        data-testid="recipes-listgroup"
      />
    </div>
    <div
      class="col-md-6"
    >
      <div>
        <br />
        <p
          data-testid="empty-recipe-details"
        >
          Please click on a Recipe
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
